qa/L0_pinned_memory/test.sh

```diff
@@ -0,0 +1,204 @@
+            echo "instance_group [ { kind: KIND_GPU }]" >> config.pbtxt)
+    
+    ENSEMBLE_NAME=${BACKEND}_ensemble
+    mkdir -p models/$ENSEMBLE_NAME/1 && \
+        cp $ENSEMBLE_NAME.pbtxt models/$ENSEMBLE_NAME/config.pbtxt
+
+    # With pinned memory
+    SERVER_ARGS="--model-repository=`pwd`/models --log-verbose=1"
+    SERVER_LOG="${ENSEMBLE_NAME}.pinned.serverlog"
+    run_server
+    if (( $SERVER_PID == 0 )); then
+        echo -e "\n***\n*** Failed to start $SERVER\n***"
+        cat $SERVER_LOG
+        exit 1
+    fi
+
+    # Sanity check that the server allocates pinned memory
+    set +e
+    $CLIENT -m${ENSEMBLE_NAME} -b1 -s1 -n1
+    if (( $? != 0 )); then
+        RET=1
+    fi
+
+    grep "] non-pinned" ${ENSEMBLE_NAME}.pinned.serverlog
+    if [ $? -eq 0 ]; then
+        echo -e "\n***\n*** Failed. Expected only pinned memory is allocated\n***"
+        RET=1
+    fi
+    set -e
+
+    kill $SERVER_PID
+    wait $SERVER_PID
+
+    # Restart the server without verbose logging
+    SERVER_ARGS="--model-repository=`pwd`/models"
+    SERVER_LOG="${ENSEMBLE_NAME}.pinned.serverlog"
+    run_server
+    if (( $SERVER_PID == 0 )); then
+        echo -e "\n***\n*** Failed to start $SERVER\n***"
+        cat $SERVER_LOG
+        exit 1
+    fi
+
+    # 16k 1m 2m 4m 8m 16m elements
+    set +e
+    for TENSOR_SIZE in 16384 1048576 2097152 4194304 8388608 16777216; do
+        $CLIENT -i grpc -u localhost:8001 -m${ENSEMBLE_NAME} \
+                -b${BATCH_SIZE} -s${TENSOR_SIZE} -n${MEASURE_ITERS} \
+                >> ${BACKEND}.${TENSOR_SIZE}.pinned.log 2>&1
+        if (( $? != 0 )); then
+            RET=1
+        fi
+    done
+    set -e
+
+    kill $SERVER_PID
+    wait $SERVER_PID
+
+    # Without pinned memory
+    SERVER_ARGS="--model-repository=`pwd`/models --pinned-memory-pool-byte-size=0 --log-verbose=1"
+    SERVER_LOG="${ENSEMBLE_NAME}.nonpinned.serverlog"
+    run_server
+    if (( $SERVER_PID == 0 )); then
+        echo -e "\n***\n*** Failed to start $SERVER\n***"
+        cat $SERVER_LOG
+        exit 1
+    fi
+
+    # Sanity check that the server allocates non-pinned memory
+    set +e
+    $CLIENT -m${ENSEMBLE_NAME} -b1 -s1 -n1
+    if (( $? != 0 )); then
+        RET=1
+    fi
+
+    grep "] pinned" ${ENSEMBLE_NAME}.nonpinned.serverlog
+    if [ $? -eq 0 ]; then
+        echo -e "\n***\n*** Failed. Expected only non-pinned memory is allocated\n***"
+        RET=1
+    fi
+    set -e
+
+    kill $SERVER_PID
+    wait $SERVER_PID
+
+    # Restart the server without verbose logging
+    SERVER_ARGS="--model-repository=`pwd`/models --pinned-memory-pool-byte-size=0"
+    SERVER_LOG="${ENSEMBLE_NAME}.nonpinned.serverlog"
+    run_server
+    if (( $SERVER_PID == 0 )); then
+        echo -e "\n***\n*** Failed to start $SERVER\n***"
+        cat $SERVER_LOG
+        exit 1
+    fi
+
+    # 16k 1m 2m 4m 8m 16m elements
+    set +e
+    for TENSOR_SIZE in 16384 1048576 2097152 4194304 8388608 16777216; do
```

Comment:
So what size do you find it to be beneficial?

Reply:
Starting from size larger than 8m elements (32 MB), the latency using pinned memory is smaller.

Reply:
Did you see this paper: https://www.cs.virginia.edu/~mwb7w/cuda_support/pinned_tradeoff.html
Your result seems to align with it. What we need to do is preallocate the entire pool of pinned memory and the allocate out chunks. So the pinned manager needs to get smarted. Can you file a 19.11 ticket for that. As for 19.10 you should just set the threshold for when pinned memory is used to be only for >32MB sizes.

Reply:
Yes, the observation of this paper makes sense to me. Should I add the threshold to this PR or a separate one.

Reply:
Separate one.